pennylane/qnn/keras.py

```diff
@@ -216,8 +216,12 @@ def __init__(
             self._signature_validation(qnode, weight_shapes)
```

Comment:
I'd like to suggest we add the following after line 187:
```rst
        **Returning a state**

        If your QNode returns the state of the quantum circuit using :func:`~.state` or
        :func:`~.density_matrix`, you must immediately follow your quantum Keras Layer with a layer
        that casts to reals. For example, you could use
        `tf.keras.layers.Lambda <https://www.tensorflow.org/api_docs/python/tf/keras/layers/Lambda>`__
        with the function ``lambda x: tf.abs(x)``. This casting is required because TensorFlow's
        Keras layers require a real input and are differentiated with respect to real parameters.
```
This would provide a warning in the documentation for users thinking to return the state or density matrix.

Reply:
> This casting is required because TensorFlow's Keras layers require a real input and are differentiated with respect to real parameters.

Sorry, but I'm unable to find the documentation where it says this. I found [this](https://www.tensorflow.org/api_docs/python/tf/dtypes/DType) and [this](https://keras.io/api/layers/core_layers/input/) which allows for all supported `dtypes` including `complex128`.

Reply:
I too am unsure, but I know that the output of a Keras layer is real even if the input is complex:
```python
import tensorflow as tf
import numpy as np

z = tf.Variable([[np.exp(1j * i * (j - 1) ** 2) for i in range(4)] for j in range(5)], dtype=tf.complex128)

with tf.GradientTape() as tape:
    res = tf.keras.layers.Dense(5)(z)

print(res)
```
This is somewhat surprising behaviour for me. I'm not sure if it's just casting the input data to real, or doing the calculation and then casting the output to real. But either way, I think users could get surprising results if they expect complex data to be passing through. Moreover, in tensorflow if we differentiate a complex tensor with respect to a real tensor, the result is real (again surprisingly). For example:
```python
import pennylane as qml
import tensorflow as tf

x = tf.Variable(0.4)

dev = qml.device("default.qubit", wires=1)

@qml.qnode(dev, interface="tf")
def f(x):
    qml.Hadamard(wires=0)
    qml.RX(x, wires=0)
    return qml.state()


with tf.GradientTape() as tape:
    res = f(x)

jac = tape.jacobian(res, x)

print(res.dtype)
print(jac.dtype)
```
The jacobian is real, unless you do `x = tf.Variable(0.4, dtype=tf.complex128)`. This has potential implications for users making a quantum Keras layer that returns a state. All the trainable weights in the quantum Keras layer are initialized as floats, so users who differentiate a complex-valued quantum Keras layer with respect to its trainable parameters will incorrectly get a real tensor (should be complex).

I guess the suggested "If your QNode returns..." was an attempt to capture this.

Reply:
Ah, got it. Then I agree that we should add the cautionary comment so that people get what they expect since everything will be real. We can work on the complex `dtype` later. I will make the appropriate changes.